### PR TITLE
feat: detect flaky tests

### LINKED ISSUE #4264: Add nightly build for detection of flaky tests
# Problem

Lately, we have been grappling with a significant number of unstable tests. Currently, pinpointing these erratic tests requires manual effort, which involves observing which tests fail upon re-execution. This approach is far from ideal.

Implementing a nightly build process that executes each test multiple times and generates an accessible report would greatly assist in detecting unstable tests. Additionally, this could help in tracing back to commits that potentially increased the instability of existing tests.

### PR DESCRIPTION
**Pull request by yakimant**
Need to create a Jenkins job to
- run tests multiple times nightly
- don't fail immediatly on test failure
- generate junit report with multiple runs
- make sure it is correctly shown in Jenkins (not done in this PR)
- the best would be to send an email report (leave for later)

Closes #4264

### CODE REVIEW

**Comment by yakimant** [comment]
Looks like `go-junit-report` is not combining the results of multiple runs.
I also could not find it in specs:
- https://github.com/testmoapp/junitxml
- https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd

As I remember I was using some software able co combine multiple runs, maybe that one:
https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/docs/feature_details/collate_junit_reports.md
And then loading the resulting xml to Teamcity succesfully.

**Comment by yakimant** [comment]
https://github.com/bhovhannes/junit-report-merger doesn't solve the issue with multirun junit report.

**Comment by yakimant** [comment]
https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/docs/feature_details/collate_junit_reports.md
provides smth better, trying to merge test results with `retries` attribute.

**Comment by yakimant** [comment]
Another option is to have our own html generation, see:
- https://github.com/vakenbolt/go-test-report (will not show failed test if it recovered)
- https://github.com/Thatooine/go-test-html-report (simmilar to jenkins handling, but worse UI)
- https://github.com/medyagh/gopogh/

And then integrate to Jenkins:
https://github.com/jenkinsci/htmlpublisher-plugin

**Comment by yakimant** [comment]
gopogh is Okish, although looks ugly:

Screenshot 2023-11-16 at 15 27 50

---

Screenshot 2023-11-16 at 15 28 12

---

Screenshot 2023-11-16 at 15 28 28

**Comment by yakimant** [comment]
For Jenkins email integration:
1. https://www.jenkins.io/doc/pipeline/steps/email-ext/
2. https://plugins.jenkins.io/mailer/

File: _assets/scripts/run_unit_tests.sh [hunk 1]
```diff
@@ -1,25 +1,45 @@
 #!/usr/bin/env bash
 set -o pipefail
+source "${GIT_ROOT}/_assets/scripts/colors.sh"
```
Comment (jakubgs) [inline _assets/scripts/run_unit_tests.sh:6]: Definition `GIT_ROOT` is below this.

File: _assets/scripts/run_unit_tests.sh [hunk 2]
```diff
@@ -1,25 +1,45 @@
...
 
 for package in ${UNIT_TEST_PACKAGES}; do
   echo -e "${GRN}Testing:${RST} ${package}"
   package_dir=$(go list -f "{{.Dir}}" "${package}")
   output_file=${package_dir}/test.log
 
-  go test -tags "${BUILD_TAGS}" -timeout 30m -v -failfast "${package}" ${GOTEST_EXTRAFLAGS} | \
-     if [ "${CI}" = "true" ]; then cat > "${output_file}"; else tee "${output_file}"; fi
+  go test -count="${UNIT_TEST_COUNT}" -tags "${BUILD_TAGS}" -v "${package}" ${GOTEST_EXTRAFLAGS} | \
+     if [[ "${CI}" == 'true' ]]; then cat > "${output_file}"; else tee "${output_file}"; fi
```
Comment (jakubgs) [inline _assets/scripts/run_unit_tests.sh]: This line looks like it could be a function.

File: _assets/scripts/test_stats.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+for file in glob.glob("**/report.xml"):
+    tree = ET.parse(file)
+    root = tree.getroot()
+    for testcase in root.iter("testcase"):
+        test_name = testcase.attrib["name"]
+
+        if test_name not in test_stats:
+            test_stats[test_name] = {"failed": 0, "total": 0}
+
+        test_stats[test_name]["total"] += 1
```
Comment (jakubgs) [inline _assets/scripts/test_stats.py]: You could use [setdefault](https://docs.python.org/2/library/stdtypes.html#dict.setdefault) for this.
```suggestion
        test_stats.setdefault(test_name, {"failed": 0, "total": 0})
        test_stats[test_name]["total"] += 1
```
But there is an even more Pythonic way using `defaultdict`:
```python
In [1]: from collections import defaultdict

In [2]: d = defaultdict(lambda: defaultdict(int))

In [3]: d['x']['z'] += 1

In [4]: d['x']['z'] += 1

In [5]: d['x']['y'] += 1

In [6]: d
Out[6]: 
defaultdict(<function __main__.<lambda>()>,
            {'x': defaultdict(int, {'z': 2, 'y': 1})})
```

File: _assets/scripts/test_stats.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+        if testcase.find("failure") is not None:
+            test_stats[test_name]["failed"] += 1
+
+failing_test_stats = [
+    {"name": name, "failure_rate": stats["failed"] / stats["total"]}
+    for name, stats in test_stats.items() if stats["failed"] != 0
+]
```
Comment (jakubgs) [inline _assets/scripts/test_stats.py:25]: Cute.

**Review by jakubgs — APPROVED** [review]
Looks simple enough.

File: _assets/scripts/test_stats.py (hunk 3 above)
Comment (yakimant) [reply _assets/scripts/test_stats.py]: Beautiful, thanks!

File: _assets/scripts/run_unit_tests.sh (hunk 2 above)
Comment (yakimant) [reply _assets/scripts/run_unit_tests.sh]: Done!

**Comment by yakimant** [comment]
10 runs took 1h, but this is not stable.
Each run cat take up to 30-40 minutes, resulting in 6h in worst case.

**Review by jakubgs — APPROVED** [review]
I'm wondering id maybe contents of the `for` loop in the `run_unit_tests.sh` could be extracted into `run_single_unit_tests.sh` or at least a function.

Not mandatory tho.

**Pull request merged**